neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,193 @@
+                 project_query=True,
+                 ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
+        self.heads = heads
+        self.dim_head = dim_head
+
+        self.project_query = project_query
+        if project_query:
+            self.to_q = nn.Linear(channels, dim_head * heads, bias=False)
+        else:
+            self.to_q = nn.Identity()
+
+        self.to_k = nn.Linear(channels, dim_head * heads, bias=False)
+
+        self.k_norm = nn.InstanceNorm1d(dim_head, affine=False)
+        self.v_norm = nn.InstanceNorm1d(dim_head, affine=False)
+
+        self.to_v = nn.Linear(channels, dim_head * heads, bias=False)
+
+        self.to_out = nn.Linear(dim_head * heads, channels) if dim_head * heads != channels else nn.Identity()
+
+        self.use_pe = use_pe
+        self.pos_dim = pos_dim
+
+        self.init_gain = 1 / math.sqrt(dim_head)
+        self.diagonal_weight = self.init_gain
+        self.initialize_qkv_weights()
+
+    def init_weight(self, weight, inif_fn):
+        # initialization for the projection matrix
+        # basically initialize the weights for each heads as
+        # this requires input dim = dim_head
+        # W = init_fn(W) + I * diagonal_weight
+        # init_fn is usually xavier_uniform_
+        for param in weight.parameters():
+            if param.ndim > 1:
+                dim_head = param.size(0) // self.heads
+                for h in range(self.heads):
+                    inif_fn(param[h * dim_head:(h + 1) * dim_head, :], gain=self.init_gain)
+
+                    param.data[h * self.dim_head:(h + 1) * self.dim_head, :] += self.diagonal_weight * \
+                                                                                    torch.diag(torch.ones(
+                                                                                    param.size(-1),
+                                                                                    dtype=torch.float32))
+
+    def initialize_qkv_weights(self):
+        init_fn = xavier_uniform_
+
+        if self.project_query:
+            self.init_weight(self.to_q, init_fn)
+        self.init_weight(self.to_k, init_fn)
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, x, norm_fn):
+        b = x.shape[0]
+        x = x.view(b*self.heads, -1, self.dim_head)
+        x = norm_fn(x)
+        return x.view(b, self.heads, -1, self.dim_head)
+
+    def forward(self,
+                u_x,
+                pos_x,
+                pos_emb=None,    # positional encoding module for encoding q/k
+                u_y=None,
+                pos_y=None,
+                weights=None,
+                associate=True,   # can be much faster if n is larger than the channel number c
+                get_kernel=False):
+        """
+        Computes kernel integral transform with attention
+
+        Parameters
+        ----------
+        u_x: input (query) function of shape [b, n, c]
```

Comment:
Can we expand b, n c to full variable name?

Reply:
Done